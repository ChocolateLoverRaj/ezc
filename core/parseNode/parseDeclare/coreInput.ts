import coreTypeParsers from '../coreTypeParsers'
import coreKeyWordsToInputFlags from '../parseInputFlags/coreKeyWordsToInputFlags'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  keyWordsToInputFlags: coreKeyWordsToInputFlags
}

export default coreInput
