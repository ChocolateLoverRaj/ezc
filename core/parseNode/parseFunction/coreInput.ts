import coreTypeParsers from '../coreTypeParsers'
import coreInstructionParsers from '../parseBlock/coreInstructionParsers'
import parseBlock from '../parseBlock/parseBlock'
import coreKeyWordsToInputFlags from '../parseInputFlags/coreKeyWordsToInputFlags'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  keyWordsToInputFlags: coreKeyWordsToInputFlags,
  parseBlock: parseBlock(coreInstructionParsers)
}

export default coreInput
