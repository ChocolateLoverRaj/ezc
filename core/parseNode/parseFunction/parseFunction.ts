import CoreTokensWithData from '../../CoreTokensWithData'
import CoreTokenType from '../../CoreTokenType'
import CoreKeyWord from '../../CoreKeyWord'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import TryParseNode from '../TryParseNode'
import Input from './Input'
import splitAsyncIterator from '../../../util/splitAsyncIterator/splitAsyncIterator'
import skipSplittedIterator from '../../../util/splitAsyncIterator/skip'
import tryNodeParsers from '../tryNodeParsers/tryNodeParsers'
import parseIdentifier from '../parseIdentifier'
import EnumItemWithData from '../../EnumItemWithData'
import parseInput from '../parseInput/parseInput'
import checkKeyWord from '../checkKeyWord'

const parseFunction = (
  { typeParsers, keyWordsToInputFlags, parseBlock }: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.FUNCTION]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
  const type = {
    enum: CoreNodeType,
    id: CoreNodeType.FUNCTION
  }

  let length = 0
  const skip = (numberOfElements: number): void => {
    skipSplittedIterator(splittedIterator, numberOfElements)
    length += numberOfElements
  }

  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected define',
      { enum: CoreKeyWord, id: CoreKeyWord.DEFINE }
    )
    if (error !== undefined) return error
    skip(1)
  }

  const parseReturnTypeResult = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
  if (!parseReturnTypeResult.success) {
    return {
      success: false,
      result: {
        type,
        index: length,
        message: 'Expected return type',
        subAttempts: parseReturnTypeResult.result
      }
    }
  }
  skip(parseReturnTypeResult.result.length)

  const parseIdentifierResult = await parseIdentifier(splittedIterator.asyncIterable)
  if (!parseIdentifierResult.success) {
    return {
      success: false,
      result: {
        type,
        index: length,
        message: 'Expected identifier',
        subAttempts: [parseIdentifierResult.result]
      }
    }
  }
  skip(parseIdentifierResult.result.length)

  // Parse (
  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected open parenthesis',
      { enum: CoreKeyWord, id: CoreKeyWord.OPEN_PARENTHESIS }
    )
    if (error !== undefined) return error
    skip(1)
  }

  // Parse inputs
  const inputs: EnumItemWithData[] = []
  for (let firstInput = true; ; firstInput = false) {
    // Check for )
    {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) {
        return {
          success: false,
          result: {
            type,
            index: length,
            message: 'Expected closing parenthesis',
            subAttempts: undefined
          }
        }
      }
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
        const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
        if (data.enum === CoreKeyWord && data.id === CoreKeyWord.CLOSE_PARENTHESIS) {
          skip(1)
          break
        }
      }
    }

    // Check for ,
    if (!firstInput) {
      const error = await checkKeyWord(
        splittedIterator.asyncIterable[Symbol.asyncIterator](),
        type,
        length,
        'Expected comma after input',
        { enum: CoreKeyWord, id: CoreKeyWord.COMMA }
      )
      if (error !== undefined) return error
      skip(1)
    }

    const parseInputInput = { typeParsers, keyWordsToInputFlags }
    const parseInputResult = await parseInput(parseInputInput)(splittedIterator.asyncIterable)
    if (!parseInputResult.success) {
      return {
        success: false,
        result: {
          type,
          index: length,
          message: 'Expected input',
          subAttempts: [parseInputResult.result]
        }
      }
    }
    skip(parseInputResult.result.length)
    inputs.push(parseInputResult.result.node)
  }

  // Parse {
  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected open curly bracket',
      { enum: CoreKeyWord, id: CoreKeyWord.OPEN_CURLY_BRACKET }
    )
    if (error !== undefined) return error
    skip(1)
  }

  // Parse blocks
  const blocks: EnumItemWithData[] = []
  const parseBlockError = await (async () => {
    while (true) {
      const parseBlockResult = await parseBlock(splittedIterator.asyncIterable)
      if (!parseBlockResult.success) return parseBlockResult.result
      skip(parseBlockResult.result.length)
      blocks.push(parseBlockResult.result.node)
    }
  })()

  // Parse }
  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected close curly bracket',
      { enum: CoreKeyWord, id: CoreKeyWord.CLOSE_CURLY_BRACKET }
    )
    if (error !== undefined) {
      return {
        success: false,
        result: {
          type,
          index: length,
          message: 'Error parsing blocks',
          subAttempts: [parseBlockError, error.result]
        }
      }
    }
    skip(1)
  }

  return {
    success: true,
    result: {
      node: {
        type,
        data: {
          name: parseIdentifierResult.result.node,
          inputs,
          returnType: parseReturnTypeResult.result.node,
          blocks
        }
      },
      length
    }
  }
}

export default parseFunction
