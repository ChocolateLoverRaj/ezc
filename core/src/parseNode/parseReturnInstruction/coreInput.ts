import coreTypeParsers from '../coreTypeParsers'
import coreValueParsers from '../coreValueParsers'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  valueParsers: coreValueParsers
}

export default coreInput
