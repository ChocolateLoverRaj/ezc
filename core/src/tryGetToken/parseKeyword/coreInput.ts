import EnumItem from '../../EnumItem'
import coreUnparseKeyWordInput from '../../unparseToken/unparseKeyWord/coreInput'
import Input from './Input'

const coreInput: Input = new Map([...coreUnparseKeyWordInput]
  .flatMap(([_enum, map]) => [...map]
    .map<[string, EnumItem]>(([id, string]) => [string, { enum: _enum, id }])))

export default coreInput
