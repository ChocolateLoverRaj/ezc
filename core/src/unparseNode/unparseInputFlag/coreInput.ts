import CoreKeyWord from '../../CoreKeyWord'
import CoreInputFlag from '../../parseNode/CoreInputFlag'
import coreKeyWordsToCoreInputFlags from
  '../../parseNode/parseInputFlags/coreKeyWordsToCoreInputFlags'
import reverseMap from 'util/dist/reverseMap/reverseMap.js'
import Input from './Input'

const coreInputFlagsToCoreKeyWords = reverseMap(coreKeyWordsToCoreInputFlags)

const coreInput: Input = new Map([
  [CoreInputFlag, new Map([...coreInputFlagsToCoreKeyWords]
    .map(([coreInputFlagId, coreKeyWordId]) => [
      coreInputFlagId,
      {
        enum: CoreKeyWord,
        id: coreKeyWordId
      }
    ]))]
])

export default coreInput
