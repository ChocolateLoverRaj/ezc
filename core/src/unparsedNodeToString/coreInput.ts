import coreUnparseTokenInput from '../unparseToken/unparseToken/coreInput'
import coreNodeFns from '../nodeFns/coreNodeFns'
import Input from './Input'

const coreInput: Input = {
  unparseTokenInput: coreUnparseTokenInput,
  allNodeFns: coreNodeFns
}

export default coreInput
