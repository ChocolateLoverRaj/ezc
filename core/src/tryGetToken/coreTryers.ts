import CoreTokenWithData from '../CoreTokenWithData'
import parseIdentifier from './parseIdentifier'
import parseIntegerType from './parseIntegerType'
import coreInput from './parseKeyword/coreInput'
import parseKeyword from './parseKeyword/parseKeyword'
import parseNumberLiteral from './parseNumberLiteral'
import parseStringLiteral from './parseStringLiteral'
import TryGetToken from './TryGetToken'

const coreTryers: ReadonlyArray<TryGetToken<CoreTokenWithData>> = [
  parseIntegerType,
  parseKeyword(coreInput),
  parseStringLiteral,
  parseIdentifier,
  parseNumberLiteral
] as const

export default coreTryers
