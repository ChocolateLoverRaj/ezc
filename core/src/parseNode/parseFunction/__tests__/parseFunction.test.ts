import arrayToAsyncIterable from 'util/dist/arrayToAsyncIterable/arrayToAsyncIterable.js'
import coreTryers from '../../../tryGetToken/coreTryers'
import parseAllTokens from '../../../tryGetToken/parseAllTokens'
import parseFunction from '../parseFunction'
import coreInput from '../coreInput'

test('simple main function', async () => {
  await expect(parseFunction(coreInput)(parseAllTokens(coreTryers)(arrayToAsyncIterable([`
    define i1 @main() {
      EntryBlock:
        ret i1 0
    }
  `])) as any)).resolves.toMatchSnapshot()
})
