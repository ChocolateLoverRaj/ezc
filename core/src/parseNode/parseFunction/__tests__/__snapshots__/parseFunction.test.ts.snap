// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple main function 1`] = `
Object {
  "result": Object {
    "length": 11,
    "node": Object {
      "data": Object {
        "blocks": Array [
          Object {
            "data": Object {
              "instructions": Array [
                Object {
                  "data": Object {
                    "type": Object {
                      "data": 1,
                      "type": Object {
                        "enum": Object {
                          "0": "FILE",
                          "1": "DECLARE",
                          "10": "IDENTIFIER",
                          "11": "NUMBER",
                          "12": "VOID_TYPE",
                          "13": "BLOCK",
                          "14": "RETURN_INSTRUCTION",
                          "15": "CALL_ASSIGNABLE",
                          "16": "ASSIGNMENT_INSTRUCTION",
                          "17": "INPUT",
                          "18": "INPUT_FLAG",
                          "2": "FUNCTION",
                          "3": "INTEGER_TYPE",
                          "4": "GLOBAL_VARIABLE",
                          "5": "STRING",
                          "6": "FLOAT_TYPE",
                          "7": "POINTER_TYPE",
                          "8": "ARRAY_TYPE",
                          "9": "STRUCT_TYPE",
                          "ARRAY_TYPE": 8,
                          "ASSIGNMENT_INSTRUCTION": 16,
                          "BLOCK": 13,
                          "CALL_ASSIGNABLE": 15,
                          "DECLARE": 1,
                          "FILE": 0,
                          "FLOAT_TYPE": 6,
                          "FUNCTION": 2,
                          "GLOBAL_VARIABLE": 4,
                          "IDENTIFIER": 10,
                          "INPUT": 17,
                          "INPUT_FLAG": 18,
                          "INTEGER_TYPE": 3,
                          "NUMBER": 11,
                          "POINTER_TYPE": 7,
                          "RETURN_INSTRUCTION": 14,
                          "STRING": 5,
                          "STRUCT_TYPE": 9,
                          "VOID_TYPE": 12,
                        },
                        "id": 3,
                      },
                    },
                    "value": Object {
                      "data": Object {
                        "floatType": undefined,
                        "value": 0,
                      },
                      "type": Object {
                        "enum": Object {
                          "0": "FILE",
                          "1": "DECLARE",
                          "10": "IDENTIFIER",
                          "11": "NUMBER",
                          "12": "VOID_TYPE",
                          "13": "BLOCK",
                          "14": "RETURN_INSTRUCTION",
                          "15": "CALL_ASSIGNABLE",
                          "16": "ASSIGNMENT_INSTRUCTION",
                          "17": "INPUT",
                          "18": "INPUT_FLAG",
                          "2": "FUNCTION",
                          "3": "INTEGER_TYPE",
                          "4": "GLOBAL_VARIABLE",
                          "5": "STRING",
                          "6": "FLOAT_TYPE",
                          "7": "POINTER_TYPE",
                          "8": "ARRAY_TYPE",
                          "9": "STRUCT_TYPE",
                          "ARRAY_TYPE": 8,
                          "ASSIGNMENT_INSTRUCTION": 16,
                          "BLOCK": 13,
                          "CALL_ASSIGNABLE": 15,
                          "DECLARE": 1,
                          "FILE": 0,
                          "FLOAT_TYPE": 6,
                          "FUNCTION": 2,
                          "GLOBAL_VARIABLE": 4,
                          "IDENTIFIER": 10,
                          "INPUT": 17,
                          "INPUT_FLAG": 18,
                          "INTEGER_TYPE": 3,
                          "NUMBER": 11,
                          "POINTER_TYPE": 7,
                          "RETURN_INSTRUCTION": 14,
                          "STRING": 5,
                          "STRUCT_TYPE": 9,
                          "VOID_TYPE": 12,
                        },
                        "id": 11,
                      },
                    },
                  },
                  "type": Object {
                    "enum": Object {
                      "0": "FILE",
                      "1": "DECLARE",
                      "10": "IDENTIFIER",
                      "11": "NUMBER",
                      "12": "VOID_TYPE",
                      "13": "BLOCK",
                      "14": "RETURN_INSTRUCTION",
                      "15": "CALL_ASSIGNABLE",
                      "16": "ASSIGNMENT_INSTRUCTION",
                      "17": "INPUT",
                      "18": "INPUT_FLAG",
                      "2": "FUNCTION",
                      "3": "INTEGER_TYPE",
                      "4": "GLOBAL_VARIABLE",
                      "5": "STRING",
                      "6": "FLOAT_TYPE",
                      "7": "POINTER_TYPE",
                      "8": "ARRAY_TYPE",
                      "9": "STRUCT_TYPE",
                      "ARRAY_TYPE": 8,
                      "ASSIGNMENT_INSTRUCTION": 16,
                      "BLOCK": 13,
                      "CALL_ASSIGNABLE": 15,
                      "DECLARE": 1,
                      "FILE": 0,
                      "FLOAT_TYPE": 6,
                      "FUNCTION": 2,
                      "GLOBAL_VARIABLE": 4,
                      "IDENTIFIER": 10,
                      "INPUT": 17,
                      "INPUT_FLAG": 18,
                      "INTEGER_TYPE": 3,
                      "NUMBER": 11,
                      "POINTER_TYPE": 7,
                      "RETURN_INSTRUCTION": 14,
                      "STRING": 5,
                      "STRUCT_TYPE": 9,
                      "VOID_TYPE": 12,
                    },
                    "id": 14,
                  },
                },
              ],
              "name": "EntryBlock",
            },
            "type": Object {
              "enum": Object {
                "0": "FILE",
                "1": "DECLARE",
                "10": "IDENTIFIER",
                "11": "NUMBER",
                "12": "VOID_TYPE",
                "13": "BLOCK",
                "14": "RETURN_INSTRUCTION",
                "15": "CALL_ASSIGNABLE",
                "16": "ASSIGNMENT_INSTRUCTION",
                "17": "INPUT",
                "18": "INPUT_FLAG",
                "2": "FUNCTION",
                "3": "INTEGER_TYPE",
                "4": "GLOBAL_VARIABLE",
                "5": "STRING",
                "6": "FLOAT_TYPE",
                "7": "POINTER_TYPE",
                "8": "ARRAY_TYPE",
                "9": "STRUCT_TYPE",
                "ARRAY_TYPE": 8,
                "ASSIGNMENT_INSTRUCTION": 16,
                "BLOCK": 13,
                "CALL_ASSIGNABLE": 15,
                "DECLARE": 1,
                "FILE": 0,
                "FLOAT_TYPE": 6,
                "FUNCTION": 2,
                "GLOBAL_VARIABLE": 4,
                "IDENTIFIER": 10,
                "INPUT": 17,
                "INPUT_FLAG": 18,
                "INTEGER_TYPE": 3,
                "NUMBER": 11,
                "POINTER_TYPE": 7,
                "RETURN_INSTRUCTION": 14,
                "STRING": 5,
                "STRUCT_TYPE": 9,
                "VOID_TYPE": 12,
              },
              "id": 13,
            },
          },
        ],
        "inputs": Array [],
        "name": Object {
          "data": Object {
            "name": "main",
            "type": 0,
          },
          "type": Object {
            "enum": Object {
              "0": "FILE",
              "1": "DECLARE",
              "10": "IDENTIFIER",
              "11": "NUMBER",
              "12": "VOID_TYPE",
              "13": "BLOCK",
              "14": "RETURN_INSTRUCTION",
              "15": "CALL_ASSIGNABLE",
              "16": "ASSIGNMENT_INSTRUCTION",
              "17": "INPUT",
              "18": "INPUT_FLAG",
              "2": "FUNCTION",
              "3": "INTEGER_TYPE",
              "4": "GLOBAL_VARIABLE",
              "5": "STRING",
              "6": "FLOAT_TYPE",
              "7": "POINTER_TYPE",
              "8": "ARRAY_TYPE",
              "9": "STRUCT_TYPE",
              "ARRAY_TYPE": 8,
              "ASSIGNMENT_INSTRUCTION": 16,
              "BLOCK": 13,
              "CALL_ASSIGNABLE": 15,
              "DECLARE": 1,
              "FILE": 0,
              "FLOAT_TYPE": 6,
              "FUNCTION": 2,
              "GLOBAL_VARIABLE": 4,
              "IDENTIFIER": 10,
              "INPUT": 17,
              "INPUT_FLAG": 18,
              "INTEGER_TYPE": 3,
              "NUMBER": 11,
              "POINTER_TYPE": 7,
              "RETURN_INSTRUCTION": 14,
              "STRING": 5,
              "STRUCT_TYPE": 9,
              "VOID_TYPE": 12,
            },
            "id": 10,
          },
        },
        "returnType": Object {
          "data": 1,
          "type": Object {
            "enum": Object {
              "0": "FILE",
              "1": "DECLARE",
              "10": "IDENTIFIER",
              "11": "NUMBER",
              "12": "VOID_TYPE",
              "13": "BLOCK",
              "14": "RETURN_INSTRUCTION",
              "15": "CALL_ASSIGNABLE",
              "16": "ASSIGNMENT_INSTRUCTION",
              "17": "INPUT",
              "18": "INPUT_FLAG",
              "2": "FUNCTION",
              "3": "INTEGER_TYPE",
              "4": "GLOBAL_VARIABLE",
              "5": "STRING",
              "6": "FLOAT_TYPE",
              "7": "POINTER_TYPE",
              "8": "ARRAY_TYPE",
              "9": "STRUCT_TYPE",
              "ARRAY_TYPE": 8,
              "ASSIGNMENT_INSTRUCTION": 16,
              "BLOCK": 13,
              "CALL_ASSIGNABLE": 15,
              "DECLARE": 1,
              "FILE": 0,
              "FLOAT_TYPE": 6,
              "FUNCTION": 2,
              "GLOBAL_VARIABLE": 4,
              "IDENTIFIER": 10,
              "INPUT": 17,
              "INPUT_FLAG": 18,
              "INTEGER_TYPE": 3,
              "NUMBER": 11,
              "POINTER_TYPE": 7,
              "RETURN_INSTRUCTION": 14,
              "STRING": 5,
              "STRUCT_TYPE": 9,
              "VOID_TYPE": 12,
            },
            "id": 3,
          },
        },
      },
      "type": Object {
        "enum": Object {
          "0": "FILE",
          "1": "DECLARE",
          "10": "IDENTIFIER",
          "11": "NUMBER",
          "12": "VOID_TYPE",
          "13": "BLOCK",
          "14": "RETURN_INSTRUCTION",
          "15": "CALL_ASSIGNABLE",
          "16": "ASSIGNMENT_INSTRUCTION",
          "17": "INPUT",
          "18": "INPUT_FLAG",
          "2": "FUNCTION",
          "3": "INTEGER_TYPE",
          "4": "GLOBAL_VARIABLE",
          "5": "STRING",
          "6": "FLOAT_TYPE",
          "7": "POINTER_TYPE",
          "8": "ARRAY_TYPE",
          "9": "STRUCT_TYPE",
          "ARRAY_TYPE": 8,
          "ASSIGNMENT_INSTRUCTION": 16,
          "BLOCK": 13,
          "CALL_ASSIGNABLE": 15,
          "DECLARE": 1,
          "FILE": 0,
          "FLOAT_TYPE": 6,
          "FUNCTION": 2,
          "GLOBAL_VARIABLE": 4,
          "IDENTIFIER": 10,
          "INPUT": 17,
          "INPUT_FLAG": 18,
          "INTEGER_TYPE": 3,
          "NUMBER": 11,
          "POINTER_TYPE": 7,
          "RETURN_INSTRUCTION": 14,
          "STRING": 5,
          "STRUCT_TYPE": 9,
          "VOID_TYPE": 12,
        },
        "id": 2,
      },
    },
  },
  "success": true,
}
`;
